addressing/README.md

```diff
@@ -0,0 +1,337 @@
+address and peer id, as described in [the section on the `p2p`
+multiaddr](#the-p2p-multiaddr).
+
+### multiaddr basics
+
+A multiaddr sequence of instructions that can be traversed to some destination.
+
+For example, the `/ip4/7.7.7.7/tcp/1234` multiaddr starts with `ip4`, which is
+the lowest-level protocol that requires an address. The `tcp` protocol runs on
+top of `ip4`, so it comes next.
+
+The multiaddr above consists of two components, the `/ip4/7.7.7.7` component,
+and the `/tcp/1234` component. It's not possible to split either one further;
+`/ip4` alone is an invalid multiaddr, because the `ip4` protocol was defined to
+require a 32 bit address. Similarly, `tcp` requires a 16 bit port number.
+
+Although we referred to `/ip4/7.7.7.7` and `/tcp/1234` as "components" of a
+larger TCP/IP address, each is actually a valid multiaddr according to the
+multiaddr spec. However, not every **syntactically valid multiaddr is a
+functional description of a process in the network**. As we've seen, even a
+simple TCP/IP connection requires composing two multiaddrs into one. See the
+section on [composing multiaddrs](#composing-multiaddrs) for information on how
+multiaddrs can be combined, and the
+[Transport multiaddrs section](#transport-multiaddrs) for the combinations that
+describe valid transport addresses.
+
+The [multiaddr protocol table][multiaddr-proto-table] contains all currently
+defined protocols and the length of their address components.
+
+### Composing multiaddrs
+
+As shown above, protocol addresses can be composed within a multiaddr in a way
+that mirrors the composition of protocols within a networking stack.
+
+The terms generally used to describe composition of multiaddrs are
+"encapsulation" and "decapsulation", and they essentially refer to adding and
+removing protocol components from a multiaddr, respectively.
+
+#### Encapsulation
+
+A protocol is said to be "encapsulated within" another protocol when data from
+an "inner" protocol is wrapped by another "outer" protocol, often by re-framing
+the data from the inner protocol into the type of packets, frames or datagrams
+used by the outer protocol.
+
+Some examples of protocol encapsulation are HTTP requests encapsulated within
+TCP/IP streams, or TCP segments themselves encapsulated within IP datagrams.
+
+The multiaddr format was designed so that addresses encapsulate each other in
+the same manner as the protocols that they describe. The result is an address
+that begins with the "outermost" layer of the network stack and works
+progressively "inward". For example, in the address `/ip4/7.7.7.7/tcp/80/ws`,
+the outermost protocol is IPv4, which encapsulates TCP streams, which in turn
+encapsulate WebSockets.
+
+All multiaddr implementations provide an `Encapsulate` method, which combines
+two multiaddrs into a composite. For example, `/ip4/7.7.7.7` can encapsulate
+`/tcp/42` to become `/ip4/7.7.7.7/tcp/42`.
```

Comment:
```suggestion
All multiaddr implementations provide a way to _encapsulate_ two multiaddrs into a composite.
For example, `/ip4/7.7.7.7` can encapsulate
`/tcp/42` to become `/ip4/7.7.7.7/tcp/42`.
```

While offering the functionality itself, the Rust multiaddr implementation does not offer an `encapsulate` method. How about this rephrased version?